File: mandel/webmap.py
from flask import Flask, request, jsonify
from pymongo import MongoClient

app = Flask(__name__)
client = MongoClient("mongodb://localhost:27017")
db = client.hashmap_db
hashmap = db.hashmap

@app.route('/<key>', methods=['GET'])
def get_value(key):
    result = hashmap.find_one({"_id": key})
    if result:
        return jsonify(value=result["value"])
    else:
        return jsonify(error="Key not found"), 404

@app.route('/<key>', methods=['PUT'])
def put_value(key):
    value = request.json.get("value")
    if value is None:
        return jsonify(error="Value is required"), 400

    hashmap.update_one({"_id": key}, {"$set": {"value": value}}, upsert=True)
    return jsonify(success=True)

@app.route('/<key>', methods=['DELETE'])
def delete_value(key):
    result = hashmap.delete_one({"_id": key})
    if result.deleted_count > 0:
        return jsonify(success=True)
    else:
        return jsonify(error="Key not found"), 404

if __name__ == '__main__':
    app.run()

File: java/src/main/java/WebMap.java
import com.google.gson.Gson;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.UpdateOptions;
import com.mongodb.client.result.DeleteResult;
import org.bson.Document;

import static com.mongodb.client.model.Filters.eq;
import static spark.Spark.*;

public class WebMap {
    public static void main(String[] args) {
        MongoClient mongoClient = MongoClients.create("mongodb://localhost:27017");
        MongoDatabase database = mongoClient.getDatabase("hashmap_db");
        MongoCollection<Document> hashmap = database.getCollection("hashmap");

        Gson gson = new Gson();

        get("/:key", (req, res) -> {
            String key = req.params(":key");
            Document result = hashmap.find(eq("_id", key)).first();
            if (result != null) {
                res.type("application/json");
                return gson.toJson(new ValueResponse(result.getString("value")));
            } else {
                res.status(404);
                return gson.toJson(new ErrorResponse("Key not found"));
            }
        });

        put("/:key", (req, res) -> {
            String key = req.params(":key");
            ValueRequest valueRequest = gson.fromJson(req.body(), ValueRequest.class);
            if (valueRequest == null || valueRequest.value == null) {
                res.status(400);
                return gson.toJson(new ErrorResponse("Value is required"));
            }

            hashmap.updateOne(eq("_id", key),
                    new Document("$set", new Document("value", valueRequest.value)), new UpdateOptions().upsert(true));
            res.type("application/json");
            return gson.toJson(new SuccessResponse(true));
        });

        delete("/:key", (req, res) -> {
            String key = req.params(":key");
            DeleteResult result = hashmap.deleteOne(eq("_id", key));
            if (result.getDeletedCount() > 0) {
                res.type("application/json");
                return gson.toJson(new SuccessResponse(true));
            } else {
                res.status(404);
                return gson.toJson(new ErrorResponse("Key not found"));
            }
        });
    }

    static class ValueRequest {
        String value;
    }

    static class ValueResponse {
        String value;

        public ValueResponse(String value) {
            this.value = value;
        }
    }

    static class SuccessResponse {
        boolean success;

        public SuccessResponse(boolean success) {
            this.success = success;
        }
    }

    static class ErrorResponse {
        String error;

        public ErrorResponse(String error) {
            this.error = error;
        }
    }
}
